text/0000-io-safety.md

```diff
@@ -0,0 +1,428 @@
+boundaries, if it isn't a valid handle or it's used after the `close` of its
+resource. And in both cases, the effects can be non-local, affecting otherwise
+unrelated parts of a program. Protection from raw pointer hazards is called
+memory safety, so protection from raw handle hazards is called *I/O safety*.
+
+Rust's standard library also has high-level types such as [`File`] and
+[`TcpStream`] which are wrappers around these raw handles, providing high-level
+interfaces to OS APIs.
+
+These high-level types also implement the traits [`FromRawFd`] on Unix-like
+platforms, and [`FromRawHandle`]/[`FromRawSocket`] on Windows, which provide
+functions which wrap a low-level value to produce a high-level value. These
+functions are unsafe, since they're unable to guarantee I/O safety. The type
+system doesn't constrain the handles passed in:
+
+```rust
+    use std::fs::File;
+    use std::os::unix::io::FromRawFd;
+
+    // Create a file.
+    let file = File::open("data.txt")?;
+
+    // Construct a `File` from an arbitrary integer value. This type checks,
+    // however 7 may not identify a live resource at runtime, or it may
+    // accidentally alias encapsulated raw handles elsewhere in the program. An
+    // `unsafe` block acknowledges that it's the caller's responsibility to
+    // avoid these hazards.
+    let forged = unsafe { File::from_raw_fd(7) };
+
+    // Obtain a copy of `file`'s inner raw handle.
+    let raw_fd = file.as_raw_fd();
+
+    // Close `file`.
+    drop(file);
+
+    // Open some unrelated file.
+    let another = File::open("another.txt")?;
+
+    // Further uses of `raw_fd`, which was `file`'s inner raw handle, would be
+    // outside the lifetime the OS associated with it. This could lead to it
+    // accidentally aliasing other otherwise encapsulated `File` instances,
+    // such as `another`. Consequently, an `unsafe` block acknowledges that
+    // it's the caller's responsibility to avoid these hazards.
+    let dangling = unsafe { File::from_raw_fd(raw_fd) };
+```
+
+Callers must ensure that the value passed into `from_raw_fd` is explicitly
+returned from the OS, and that `from_raw_fd`'s return value won't outlive the
+lifetime the OS associates with the handle.
+
+I/O safety is new as an explicit concept, but it reflects common practices.
+Rust's `std` will require no changes to stable interfaces, beyond the
+introduction of some new types and traits and new impls for them. Initially,
+not all of the Rust ecosystem will support I/O safety though; adoption will
+be gradual.
+
+## `OwnedFd` and `BorrowedFd<'owned>`
+
+These two types are conceptual replacements for `RawFd`, and represent owned
+and borrowed handle values. `OwnedFd` owns a file descriptor, including closing
+it when it's dropped. `BorrowedFd`'s lifetime parameter ties it to the lifetime
+of something that owns a file descriptor. These types enforce all of their I/O
```

Comment:
I would phrase it as

> `BorrowedFd`'s lifetime parameter says for how long access to this file descriptor has been borrowed.

Just like with `&'a T`, this lifetime may be an under-approximation of the actual lifetime of the underlying object, and that's okay. We often borrow things for less than the full lifetime of an object.

Do you consider these two formulations to be equivalent? They sound non-equivalent to me, so I wonder what lead you to your choice.

Reply:
You're right. I was just thinking that it's obvious that a user could always drop a `BorrowedFd` earlier than needed, but it's better to clarify. I've now updated it to use your proposed wording.